### PR TITLE
Bug fix: Batch dot

### PR DESCRIPTION
**Pull request by farizrahman4u**
* Almost total rewrite of `batch_dot` for Tensorflow and CNTK backend.
* Fixes various bugs in edge cases(e.g #11035)
* Rewrite `batch_dot` implementation in `reference_operations.py`  (had bugs and overall implementation was bloated)
* Does early input validation with helpful messages.
* Earlier, the methods had partial `tf.tensordot()` like behavior, which was undocumented. This was possibly a bug first introduced in TF backend and the logic was probably copied when writing the CNTK backend. This now returns with a helpful error message.
* Reduces number of ops as much as possible.
* Slightly better docstrings
* Thorough unit tests.

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for taking care of this, it had been outstanding for a long time. To be clear, you are normalizing on the Theano behavior?

**Comment by farizrahman4u** [comment]
> To be clear, you are normalizing on the Theano behavior?

Yes.

File: tests/keras/backend/backend_test.py [hunk 1]
```diff
@@ -248,21 +258,55 @@ def test_random_variables(self):
...
-        xy_batch_dot = K.batch_dot(x_batch, y_batch, axes=0)
-        assert_allclose(K.eval(xy_batch_dot), np.ones((20, 1)) * 32, atol=1e-05)
-        # making sure swapping axes when ndim == 2 works
-        x_batch = K.ones(shape=(32, 20))
-        y_batch = K.ones(shape=(20, 32))
-        xy_batch_dot = K.batch_dot(x_batch, y_batch, axes=(0, 1))
-        assert_allclose(K.eval(xy_batch_dot), np.ones((20, 1)) * 32, atol=1e-05)
-        xy_batch_dot = K.batch_dot(x_batch, y_batch, axes=(1, 0))
-        assert_allclose(K.eval(xy_batch_dot), np.ones((32, 1)) * 20, atol=1e-05)
+        # Note : batch_dot behavior is different for
```
Comment (farizrahman4u) [inline tests/keras/backend/backend_test.py]: "implementation", not "behavior"

File: tests/keras/backend/backend_test.py [hunk 2]
```diff
@@ -248,21 +258,55 @@ def test_random_variables(self):
...
+        test_cases.append([(None, 4), (None, 3, 4), (1, 2), (None, 3)])
+        test_cases.append([(None, 4), (None, 4), None, (None, 1)])
+
+        batch_size = 7
+
+        def batch_shape(shape):
+            return (batch_size, ) + shape[1:]
+
+        def ones(shape):
+            return np.ones(batch_shape(shape))
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/backend_test.py]: Maybe should we use `np.random.uniform` and compare with the output of the `batch_dot` in `reference_operations.py` rather than hard coding the solution?

File: tests/keras/backend/backend_test.py [hunk 3]
```diff
@@ -248,21 +258,55 @@ def test_random_variables(self):
...
+        # Note : batch_dot implementation is different for
+        # placeholders and variables in CNTK backend
+
+        test_cases = []
+        test_cases.append([(None, 3, 4, 5), (None, 2, 3, 4), (2, 3),
+                           (None, 3, 5, 2, 3)])
+        test_cases.append([(None, 3, 4, 5), (None, 2, 4), 2, (None, 3, 5, 2)])
+        test_cases.append([(None, 3, 4), (None, 2, 3, 4), (2, 3), (None, 3, 2, 3)])
+        test_cases.append([(None, 4), (None, 3, 4), (1, 2), (None, 3)])
+        test_cases.append([(None, 4), (None, 4), None, (None, 1)])
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/backend_test.py]: We could add `test_cases.append([(None, 4, 3), (None, 3, 5), (2, 1), (None, 4, 5)])` to reproduce the bug report of #11035 ?

File: tests/keras/backend/backend_test.py [hunk 4]
```diff
@@ -248,21 +258,55 @@ def test_random_variables(self):
...
+
+        batch_size = 7
+
+        def batch_shape(shape):
+            return (batch_size, ) + shape[1:]
+
+        def ones(shape):
+            return np.ones(batch_shape(shape))
+
+        for tc in test_cases:
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/backend_test.py]: I suggest `for x_shape, y_shape, axes, z_shape in test_cases:` to improve readability.

File: keras/backend/cntk_backend.py [hunk 5]
```diff
@@ -1133,6 +1232,11 @@ def concatenate(tensors, axis=-1):
     return C.splice(*tensors, axis=axis[0])
 
 
+def stack(tensors, axis=0):
```
Comment (gabrieldemarmiesse) [inline keras/backend/cntk_backend.py]: you should use `def stack(x, axis=0):` to be consistent with other backends.

**Comment by farizrahman4u** [comment]
@gabrieldemarmiesse Done.

File: keras/backend/cntk_backend.py [hunk 6]
```diff
@@ -570,43 +570,142 @@ def batch_dot(x, y, axes=None):
...
+
+    # convert negative indices
+    if axes[0] < 0:
+        axes[0] += x_ndim
+    if axes[1] < 0:
+        axes[1] += y_ndim
+
+    if 0 in axes:
+        raise ValueError('Can not perform batch_dot over axis 0.'
+                         'If your inputs are not batched,'
```
Comment (bdwyer2) [inline keras/backend/cntk_backend.py]: Forgot a space
```suggestion
                         ' If your inputs are not batched,'
```

**Comment by fchollet** [comment]
Incidentally, could you please add a unit test checking that `Dot._compute_output_shape()` has the correct behavior across the various input shapes & axes that we need to support?

**Comment by farizrahman4u** [comment]
@fchollet  Will do in separate PR.

**Comment by fchollet** [comment]
@farizrahman4u if you have the bandwidth, it would also be very helpful to apply the same fix in tf.keras in the TensorFlow repo.

**Comment by farizrahman4u** [comment]
@fchollet https://github.com/tensorflow/tensorflow/pull/23305

The 2 space indentation though :(

**Comment by gabrieldemarmiesse** [comment]
> The 2 space indentation though :(

Maybe that's why tf.keras couldn't be a simple fork :D
